File: articles/r-chemistry-statistics-experimental-analysis/python/replicate_summary.py
"""
Replicate summary scaffold for chemistry statistics.

Run from article directory:
    python python/replicate_summary.py
"""

from pathlib import Path
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
INPUT_PATH = ARTICLE_DIR / "data" / "replicate_measurements.csv"
OUTPUT_PATH = ARTICLE_DIR / "outputs" / "tables" / "python_replicate_summary.csv"


def main() -> None:
    OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)

    data = pd.read_csv(INPUT_PATH)
    summary = (
        data.groupby(["sample_id", "analyte", "method_id", "batch_id"], as_index=False)
        .agg(
            mean_mM=("measurement_mM", "mean"),
            sd_mM=("measurement_mM", "std"),
            n=("measurement_mM", "count"),
        )
    )
    summary["se_mM"] = summary["sd_mM"] / summary["n"] ** 0.5
    summary["rsd_percent"] = 100 * summary["sd_mM"] / summary["mean_mM"]

    summary.to_csv(OUTPUT_PATH, index=False)
    print(summary.round(6).to_string(index=False))
    print(f"Saved: {OUTPUT_PATH}")


if __name__ == "__main__":
    main()

File: articles/python-chemistry-simulation-laboratory-data/r/replicate_summary.R
# Replicate summary scaffold.
# Synthetic educational data only.

measurements <- read.csv(file.path("data", "replicate_measurements.csv"))

mean_table <- aggregate(measurement_mM ~ sample_id + method_id, data = measurements, mean)
sd_table <- aggregate(measurement_mM ~ sample_id + method_id, data = measurements, sd)
n_table <- aggregate(measurement_mM ~ sample_id + method_id, data = measurements, length)

summary_table <- merge(mean_table, sd_table, by = c("sample_id", "method_id"), suffixes = c("_mean", "_sd"))
summary_table <- merge(summary_table, n_table, by = c("sample_id", "method_id"))
names(summary_table)[names(summary_table) == "measurement_mM"] <- "n"

summary_table$se_mM <- summary_table$measurement_mM_sd / sqrt(summary_table$n)

print(summary_table)
